File: Day 23 - Binary Tree/Zig-zag tree traversal.java
class GFG
{
	public ArrayList<Integer> zigZagTraversal(Node root) {
        ArrayList<Integer> ans = new ArrayList<>();
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        boolean leftToRight = true;
        while (!q.isEmpty()) {
            ArrayList<Integer> cur = new ArrayList<>();
            int sz = q.size();
            while (sz-- > 0) {
                Node n = q.poll();
                cur.add(n.data);
                if (n.left != null) q.add(n.left);
                if (n.right != null) q.add(n.right);
            }
            if (!leftToRight) Collections.reverse(cur);
            ans.addAll(cur);
            leftToRight ^= true;
        }
        return ans;
    }
}

File: Day 23 - Binary Tree/Zig-zag tree traversal.cpp
class Solution
{
public:
  // Function to store the zig zag order traversal of tree in a list.
  vector<int> zigZagTraversal(Node *root)
  {
    vector<int> ans;
    queue<Node *> q;
    q.push(root);
    bool leftToRight = 1;
    while (!q.empty())
    {
      vector<int> cur;
      int sz = q.size();
      while (sz--)
      {
        Node *n = q.front();
        q.pop();
        cur.push_back(n->data);
        if (n->left != NULL)
          q.push(n->left);
        if (n->right != NULL)
          q.push(n->right);
      }
      if (leftToRight == 0)
        reverse(cur.begin(), cur.end());
      for (int x : cur)
      {
        ans.push_back(x);
      }
      leftToRight ^= 1;
    }
    return ans;
  }
};